src/meta/src/barrier/mod.rs

```diff
@@ -158,6 +171,51 @@ pub struct GlobalBarrierManager<S: MetaStore> {
     metrics: Arc<MetaMetrics>,
 
     env: MetaSrvEnv<S>,
+
+    epoch_heap: Arc<RwLock<BinaryHeap<HeapNode>>>,
+
+    wait_recovery: Arc<AtomicBool>,
+
+    wait_build_actor: Arc<AtomicBool>,
+}
+
+struct HeapNode {
+    prev_epoch: u64,
+    tx: Option<oneshot::Sender<()>>,
+}
+
+impl HeapNode {
+    fn new(prev_epoch: u64, tx: Option<oneshot::Sender<()>>) -> Self {
+        Self { prev_epoch, tx }
+    }
+
+    fn notify_wait(&mut self) {
+        if let Some(tx) = self.tx.take() {
+            tx.send(()).ok();
+        }
+    }
+}
+
+impl Eq for HeapNode {}
+impl PartialEq for HeapNode {
+    fn eq(&self, other: &Self) -> bool {
+        other.prev_epoch == self.prev_epoch
+    }
+}
+
+impl PartialOrd for HeapNode {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        other.prev_epoch.partial_cmp(&self.prev_epoch)
+    }
+}
+
+impl Ord for HeapNode
+where
+    Self: PartialOrd,
+{
```

Comment:
By convention, we may write the comparison logic in `impl Ord` and implement `PartialOrd` by
```rust
fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
    Some(self.cmp(other))
}
```

Reply:
Ok, I will modify him